connection panic When close

#### Expected behavior

no panic

#### Actual behavior

panic

#### Steps to reproduce

``` go
panic: sync: WaitGroup is reused before previous Wait has returned

goroutine 129942961 gp=0xc0127e4540 m=3 mp=0xc00006d008 [running]:
panic({0x12d3ae0?, 0x1792360?})
        /home/jenkins/go1.25/src/runtime/panic.go:802 +0x168 fp=0xc002f1fe50 sp=0xc002f1fda0 pc=0x48ea28
sync.(*WaitGroup).Wait(0xc02b974378)
        /home/jenkins/go1.25/src/sync/waitgroup.go:208 +0xf4 fp=0xc002f1fe78 sp=0xc002f1fe50 pc=0x4b01b4
github.com/apache/pulsar-client-go/pulsar/internal.(*connection).failLeftRequestsWhenClose(0xc02b9742c0)
        /home/jenkins/go/pkg/mod/github.com/apache/pulsar-client-go@v0.17.0/pulsar/internal/connection.go:378 +0x25 fp=0xc002f1feb8 sp=0xc002f1fe78 pc=0xc0c745
github.com/apache/pulsar-client-go/pulsar/internal.(*connection).run(0xc02b9742c0)
        /home/jenkins/go/pkg/mod/github.com/apache/pulsar-client-go@v0.17.0/pulsar/internal/connection.go:430 +0x3bf fp=0xc002f1ffc0 sp=0xc002f1feb8 pc=0xc0cc7f
github.com/apache/pulsar-client-go/pulsar/internal.(*connection).start.func1()
        /home/jenkins/go/pkg/mod/github.com/apache/pulsar-client-go@v0.17.0/pulsar/internal/connection.go:240 +0x65 fp=0xc002f1ffe0 sp=0xc002f1ffc0 pc=0xc0ba65
runtime.goexit({})
        /home/jenkins/go1.25/src/runtime/asm_amd64.s:1693 +0x1 fp=0xc002f1ffe8 sp=0xc002f1ffe0 pc=0x4976a1
created by github.com/apache/pulsar-client-go/pulsar/internal.(*connection).start in goroutine 129886168
        /home/jenkins/go/pkg/mod/github.com/apache/pulsar-client-go@v0.17.0/pulsar/internal/connection.go:236 +0x68
```
close connection  while sending data demo code like this:
``` go
package main

import (
	"fmt"
	"sync"
	"time"
)

func simulateConnectionFuncCall(wg *sync.WaitGroup)  {
	wg.Add(1)
	time.Sleep(100 * time.Millisecond)
	wg.Done() 
}

func demoPanic(){
	wg := &sync.WaitGroup{}
	// one goroutine send something
	go func(wg *sync.WaitGroup) {
		for j:=0;j< 100;j++{
			simulateConnectionFuncCall(wg)
        }
	}(wg)


	time.Sleep(200*time.Millisecond)
	// sync.(*WaitGroup).Wait panic（waitgroup.go:208）
	wg.Wait()
	fmt.Printf("done")
}
func main() {
	//demo()
	demoPanic()
	return
}
```

#### System configuration
**Pulsar version**:  2.6.x
sdk version: 0.17.0

--- a/pulsar/internal/connection.go
+++ b/pulsar/internal/connection.go
@@ -373,31 +373,21 @@ func (c *connection) waitUntilReady() error {
 }
 
 func (c *connection) failLeftRequestsWhenClose() {
-	// wait for outstanding incoming requests to complete before draining
-	// and closing the channel
 	c.incomingRequestsWG.Wait()
 
-	ch := c.incomingRequestsCh
-	go func() {
-		// send a nil message to drain instead of
-		// closing the channel and causing a potential panic
-		//
-		// if other requests come in after the nil message
-		// then the RPC client will time out
-		ch <- nil
-		c.writeRequestsCh <- nil
-	}()
-	for req := range ch {
-		if nil == req {
-			break // we have drained the requests
+	for {
+		select {
+		case req := <-c.incomingRequestsCh:
+			if req != nil && req.callback != nil {
+				req.callback(req.cmd, ErrConnectionClosed)
+			}
+		case req := <-c.writeRequestsCh:
+			if req != nil {
+				req.data.Release()
+			}
+		default:
+			return
 		}
-		c.internalSendRequest(req)
-	}
-	for req := range c.writeRequestsCh {
-		if nil == req {
-			break
-		}
-		req.data.Release()
 	}
 }
 
@@ -656,33 +646,37 @@ func (c *connection) checkServerError(err *pb.ServerError) {
 
 func (c *connection) SendRequest(requestID uint64, req *pb.BaseCommand,
 	callback func(command *pb.BaseCommand, err error)) {
+	c.mu.RLock()
+	if c.getState() == connectionClosed {
+		c.mu.RUnlock()
+		callback(req, ErrConnectionClosed)
+		return
+	}
 	c.incomingRequestsWG.Add(1)
+	c.mu.RUnlock()
 	defer c.incomingRequestsWG.Done()
 
-	if c.getState() == connectionClosed {
+	select {
+	case <-c.closeCh:
 		callback(req, ErrConnectionClosed)
 
-	} else {
-		select {
-		case <-c.closeCh:
-			callback(req, ErrConnectionClosed)
-
-		case c.incomingRequestsCh <- &request{
-			id:       &requestID,
-			cmd:      req,
-			callback: callback,
-		}:
-		}
+	case c.incomingRequestsCh <- &request{
+		id:       &requestID,
+		cmd:      req,
+		callback: callback,
+	}:
 	}
 }
 
 func (c *connection) SendRequestNoWait(req *pb.BaseCommand) error {
-	c.incomingRequestsWG.Add(1)
-	defer c.incomingRequestsWG.Done()
-
+	c.mu.RLock()
 	if c.getState() == connectionClosed {
+		c.mu.RUnlock()
 		return ErrConnectionClosed
 	}
+	c.incomingRequestsWG.Add(1)
+	c.mu.RUnlock()
+	defer c.incomingRequestsWG.Done()
 
 	select {
 	case <-c.closeCh:

--- a/pulsar/internal/connection.go
+++ b/pulsar/internal/connection.go
@@ -455,6 +455,13 @@ func (c *connection) runPingCheck(pingCheckTicker *time.Ticker) {
 }
 
 func (c *connection) WriteData(ctx context.Context, data Buffer) {
+	if !c.registerIncomingRequest() {
+		data.Release()
+		c.log.Debug("Write data connection closed")
+		return
+	}
+	defer c.incomingRequestsWG.Done()
+
 	writeToQueue := false
 	defer func() {
 		if !writeToQueue {
@@ -644,16 +651,24 @@ func (c *connection) checkServerError(err *pb.ServerError) {
 	}
 }
 
+func (c *connection) registerIncomingRequest() bool {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	if c.getState() == connectionClosed {
+		return false
+	}
+
+	c.incomingRequestsWG.Add(1)
+	return true
+}
+
 func (c *connection) SendRequest(requestID uint64, req *pb.BaseCommand,
 	callback func(command *pb.BaseCommand, err error)) {
-	c.mu.RLock()
-	if c.getState() == connectionClosed {
-		c.mu.RUnlock()
+	if !c.registerIncomingRequest() {
 		callback(req, ErrConnectionClosed)
 		return
 	}
-	c.incomingRequestsWG.Add(1)
-	c.mu.RUnlock()
 	defer c.incomingRequestsWG.Done()
 
 	select {
@@ -669,13 +684,9 @@ func (c *connection) SendRequest(requestID uint64, req *pb.BaseCommand,
 }
 
 func (c *connection) SendRequestNoWait(req *pb.BaseCommand) error {
-	c.mu.RLock()
-	if c.getState() == connectionClosed {
-		c.mu.RUnlock()
+	if !c.registerIncomingRequest() {
 		return ErrConnectionClosed
 	}
-	c.incomingRequestsWG.Add(1)
-	c.mu.RUnlock()
 	defer c.incomingRequestsWG.Done()
 
 	select {

--- a/pulsar/internal/connection_test.go
+++ b/pulsar/internal/connection_test.go
@@ -1,0 +1,48 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package internal
+
+import (
+	"context"
+	"sync/atomic"
+	"testing"
+
+	"github.com/apache/pulsar-client-go/pulsar/log"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestConnectionWriteDataShouldNotEnqueueWhenStateClosed(t *testing.T) {
+	released := atomic.Int64{}
+	pool := NewBufferPool()
+	buf := pool.GetBuffer(8)
+	buf.SetReleaseCallback(func() {
+		released.Add(1)
+	})
+
+	c := &connection{
+		log:             log.DefaultNopLogger(),
+		closeCh:         make(chan struct{}),
+		writeRequestsCh: make(chan *dataRequest, 1),
+	}
+	c.setStateClosed()
+
+	c.WriteData(context.Background(), buf)
+
+	assert.Equal(t, 0, len(c.writeRequestsCh))
+	assert.EqualValues(t, 1, released.Load())
+}

--- a/pulsar/internal/connection_test.go
+++ b/pulsar/internal/connection_test.go
@@ -21,6 +21,7 @@ import (
 	"context"
 	"net/url"
 	"sync"
+	"sync/atomic"
 	"testing"
 	"time"
 
@@ -42,20 +43,20 @@ func TestConnectionRejectRequestsAfterClose(t *testing.T) {
 func TestConnectionSendRequestRaceWithClose(t *testing.T) {
 	// Regression test for concurrent Add/Wait on WaitGroup during Close.
 	//
-	// Without proper synchronization between:
-	//   - registerIncomingRequest() calling WaitGroup.Add(1)
-	//   - Close() calling WaitGroup.Wait()
+	// Without proper synchronization in registerIncomingRequest(), calling
+	// WaitGroup.Add(1) and checking state under c.mu.RLock(), a concurrent
+	// failLeftRequestsWhenClose() calling WaitGroup.Wait() could race with Add()
+	// in Go 1.25+, causing panic: "sync: WaitGroup is reused before previous Wait has returned"
 	//
-	// Go 1.25+ may panic with:
-	//
-	//   sync: WaitGroup is reused before previous Wait has returned
-	//
-	// This test continuously issues requests while concurrently closing
-	// the connection to maximize the Add/Wait overlap window.
+	// This test directly exercises the synchronization:
+	// 1. Many goroutines call registerIncomingRequest() to Add() to the WaitGroup
+	// 2. After they exit via Done(), failLeftRequestsWhenClose() calls Wait()
+	// 3. The test also verifies that all three methods (SendRequest, SendRequestNoWait, WriteData)
+	//    properly use registerIncomingRequest() and reject calls after close
 
 	const (
-		numTrials     = 20
-		numGoroutines = 100
+		numTrials     = 10
+		numGoroutines = 50
 	)
 
 	for trial := 0; trial < numTrials; trial++ {
@@ -63,17 +64,17 @@ func TestConnectionSendRequestRaceWithClose(t *testing.T) {
 
 		startCh := make(chan struct{})
 		stopCh := make(chan struct{})
-
-		panicCh := make(chan any, numGoroutines)
+		panicCh := make(chan any, 1)
 
 		var wg sync.WaitGroup
+		var registerCount int32
 
+		// Producer goroutines that register requests
 		for i := 0; i < numGoroutines; i++ {
 			wg.Add(1)
 
-			go func(idx int) {
+			go func() {
 				defer wg.Done()
-
 				defer func() {
 					if r := recover(); r != nil {
 						panicCh <- r
@@ -89,38 +90,60 @@ func TestConnectionSendRequestRaceWithClose(t *testing.T) {
 					default:
 					}
 
-					if idx%2 == 0 {
-						c.SendRequest(
-							uint64(idx),
-							&pb.BaseCommand{},
-							func(*pb.BaseCommand, error) {},
-						)
-					} else {
-						_ = c.SendRequestNoWait(&pb.BaseCommand{})
+					// Call registerIncomingRequest() directly to exercise the WaitGroup Add/state check
+					if c.registerIncomingRequest() {
+						atomic.AddInt32(&registerCount, 1)
+						c.incomingRequestsWG.Done()
 					}
 				}
-			}(i)
+			}()
 		}
 
-		// Start all concurrent request producers together.
+		// Start producers
 		close(startCh)
 
-		// Allow requests to race with Close().
-		time.Sleep(10 * time.Millisecond)
+		// Let producers run and accumulate adds
+		time.Sleep(20 * time.Millisecond)
 
-		c.Close()
+		// Close the connection state
+		c.mu.Lock()
+		c.setStateClosed()
+		c.mu.Unlock()
 
+		// Signal producers to stop
 		close(stopCh)
 
+		// Wait for all producers to finish
 		wg.Wait()
 
+		// Now call failLeftRequestsWhenClose() which calls Wait() on the WaitGroup
+		// With the race fixed, this should complete without panic
+		drainDone := make(chan struct{})
+		go func() {
+			defer func() {
+				if r := recover(); r != nil {
+					panicCh <- r
+				}
+			}()
+			c.failLeftRequestsWhenClose()
+			close(drainDone)
+		}()
+
+		// Wait for drain to complete
+		select {
+		case <-drainDone:
+		case <-time.After(5 * time.Second):
+			t.Fatal("failLeftRequestsWhenClose() did not complete (deadlock in WaitGroup)")
+		}
+
+		// Check for panic
 		select {
 		case p := <-panicCh:
-			t.Fatalf("unexpected panic during concurrent Close: %v", p)
+			t.Fatalf("trial %d: panic during WaitGroup race: %v", trial, p)
 		default:
 		}
 
-		assertConnectionClosed(t, c)
+		t.Logf("trial %d: %d successful registers", trial, atomic.LoadInt32(&registerCount))
 	}
 }
 

--- a/pulsar/internal/connection_test.go
+++ b/pulsar/internal/connection_test.go
@@ -155,7 +155,7 @@ func assertConnectionClosed(t *testing.T, c *connection) {
 	c.SendRequest(
 		999,
 		&pb.BaseCommand{},
-		func(cmd *pb.BaseCommand, err error) {
+		func(_ *pb.BaseCommand, err error) {
 			callbackCh <- err
 		},
 	)

--- a/pulsar/internal/connection_test.go
+++ b/pulsar/internal/connection_test.go
@@ -50,9 +50,10 @@ func TestConnectionSendRequestRaceWithClose(t *testing.T) {
 	//
 	// This test directly exercises the synchronization:
 	// 1. Many goroutines call registerIncomingRequest() to Add() to the WaitGroup
-	// 2. After they exit via Done(), failLeftRequestsWhenClose() calls Wait()
-	// 3. The test also verifies that all three methods (SendRequest, SendRequestNoWait, WriteData)
-	//    properly use registerIncomingRequest() and reject calls after close
+	// 2. While they are still running, failLeftRequestsWhenClose() calls Wait()
+	// 3. The connection transitions to closed so new registrations are rejected
+	//    and existing ones drain, letting Wait() return
+	// 4. The test verifies no panic occurs during the Add/Wait overlap
 
 	const (
 		numTrials     = 10
@@ -102,22 +103,23 @@ func TestConnectionSendRequestRaceWithClose(t *testing.T) {
 		// Start producers
 		close(startCh)
 
-		// Let producers run and accumulate adds
+		// Let producers run and accumulate pending adds
 		time.Sleep(20 * time.Millisecond)
 
-		// Close the connection state
+		// Transition the connection to closed — this runs under the write lock,
+		// matching the real Close() flow. After this, registerIncomingRequest()
+		// will reject new Add() calls, but goroutines already past the state
+		// check and holding RLock will still complete their Add()/Done().
 		c.mu.Lock()
 		c.setStateClosed()
 		c.mu.Unlock()
 
-		// Signal producers to stop
-		close(stopCh)
-
-		// Wait for all producers to finish
-		wg.Wait()
-
-		// Now call failLeftRequestsWhenClose() which calls Wait() on the WaitGroup
-		// With the race fixed, this should complete without panic
+		// Immediately start failLeftRequestsWhenClose() in a goroutine — it
+		// calls Wait(). With the fix, goroutines that already called Add()
+		// under RLock will finish their Done(), and no new Add() can happen
+		// because setStateClosed() above drained pending RLock holders. Without
+		// the fix, a goroutine slipping through could call Add() after Wait()
+		// returns, causing "WaitGroup is reused before previous Wait has returned".
 		drainDone := make(chan struct{})
 		go func() {
 			defer func() {
@@ -129,12 +131,18 @@ func TestConnectionSendRequestRaceWithClose(t *testing.T) {
 			close(drainDone)
 		}()
 
+		// Signal producers to stop
+		close(stopCh)
+
 		// Wait for drain to complete
 		select {
 		case <-drainDone:
 		case <-time.After(5 * time.Second):
 			t.Fatal("failLeftRequestsWhenClose() did not complete (deadlock in WaitGroup)")
 		}
+
+		// Wait for all producers to finish (they should already be done)
+		wg.Wait()
 
 		// Check for panic
 		select {

--- a/pulsar/internal/connection_test.go
+++ b/pulsar/internal/connection_test.go
@@ -19,30 +19,177 @@ package internal
 
 import (
 	"context"
-	"sync/atomic"
+	"net/url"
+	"sync"
 	"testing"
+	"time"
 
+	pb "github.com/apache/pulsar-client-go/pulsar/internal/pulsar_proto"
 	"github.com/apache/pulsar-client-go/pulsar/log"
+	"github.com/prometheus/client_golang/prometheus"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
-func TestConnectionWriteDataShouldNotEnqueueWhenStateClosed(t *testing.T) {
-	released := atomic.Int64{}
-	pool := NewBufferPool()
-	buf := pool.GetBuffer(8)
-	buf.SetReleaseCallback(func() {
-		released.Add(1)
-	})
+func TestConnectionRejectRequestsAfterClose(t *testing.T) {
+	c := newTestConnection()
 
-	c := &connection{
-		log:             log.DefaultNopLogger(),
-		closeCh:         make(chan struct{}),
-		writeRequestsCh: make(chan *dataRequest, 1),
+	c.Close()
+
+	assertConnectionClosed(t, c)
+}
+
+func TestConnectionSendRequestRaceWithClose(t *testing.T) {
+	// Regression test for concurrent Add/Wait on WaitGroup during Close.
+	//
+	// Without proper synchronization between:
+	//   - registerIncomingRequest() calling WaitGroup.Add(1)
+	//   - Close() calling WaitGroup.Wait()
+	//
+	// Go 1.25+ may panic with:
+	//
+	//   sync: WaitGroup is reused before previous Wait has returned
+	//
+	// This test continuously issues requests while concurrently closing
+	// the connection to maximize the Add/Wait overlap window.
+
+	const (
+		numTrials     = 20
+		numGoroutines = 100
+	)
+
+	for trial := 0; trial < numTrials; trial++ {
+		c := newTestConnection()
+
+		startCh := make(chan struct{})
+		stopCh := make(chan struct{})
+
+		panicCh := make(chan any, numGoroutines)
+
+		var wg sync.WaitGroup
+
+		for i := 0; i < numGoroutines; i++ {
+			wg.Add(1)
+
+			go func(idx int) {
+				defer wg.Done()
+
+				defer func() {
+					if r := recover(); r != nil {
+						panicCh <- r
+					}
+				}()
+
+				<-startCh
+
+				for {
+					select {
+					case <-stopCh:
+						return
+					default:
+					}
+
+					if idx%2 == 0 {
+						c.SendRequest(
+							uint64(idx),
+							&pb.BaseCommand{},
+							func(*pb.BaseCommand, error) {},
+						)
+					} else {
+						_ = c.SendRequestNoWait(&pb.BaseCommand{})
+					}
+				}
+			}(i)
+		}
+
+		// Start all concurrent request producers together.
+		close(startCh)
+
+		// Allow requests to race with Close().
+		time.Sleep(10 * time.Millisecond)
+
+		c.Close()
+
+		close(stopCh)
+
+		wg.Wait()
+
+		select {
+		case p := <-panicCh:
+			t.Fatalf("unexpected panic during concurrent Close: %v", p)
+		default:
+		}
+
+		assertConnectionClosed(t, c)
 	}
-	c.setStateClosed()
+}
+
+func assertConnectionClosed(t *testing.T, c *connection) {
+	t.Helper()
+
+	callbackCh := make(chan error, 1)
+
+	c.SendRequest(
+		999,
+		&pb.BaseCommand{},
+		func(cmd *pb.BaseCommand, err error) {
+			callbackCh <- err
+		},
+	)
+
+	select {
+	case err := <-callbackCh:
+		assert.Error(t, err)
+	case <-time.After(time.Second):
+		t.Fatal("SendRequest callback was not invoked")
+	}
+
+	assert.Error(t, c.SendRequestNoWait(&pb.BaseCommand{}))
+
+	released := make(chan struct{}, 1)
+
+	buf := NewBufferPool().GetBuffer(8)
+	buf.SetReleaseCallback(func() {
+		released <- struct{}{}
+	})
 
 	c.WriteData(context.Background(), buf)
 
-	assert.Equal(t, 0, len(c.writeRequestsCh))
-	assert.EqualValues(t, 1, released.Load())
+	select {
+	case <-released:
+	case <-time.After(time.Second):
+		t.Fatal("WriteData buffer was not released")
+	}
+}
+
+func newTestConnection() *connection {
+	opts := connectionOptions{
+		logicalAddr:       &url.URL{Host: "test:6650"},
+		physicalAddr:      &url.URL{Host: "test:6650"},
+		connectionTimeout: time.Second,
+		keepAliveInterval: 30 * time.Second,
+		logger:            log.DefaultNopLogger(),
+		metrics:           newMockMetrics(),
+	}
+
+	c := newConnection(opts)
+
+	require.NotNil(&testing.T{}, c)
+
+	return c
+}
+
+// newMockMetrics creates Metrics with real prometheus counters for testing.
+func newMockMetrics() *Metrics {
+	return &Metrics{
+		ConnectionsClosed: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "test_connections_closed",
+		}),
+		ConnectionsEstablishmentErrors: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "test_connections_establishment_errors",
+		}),
+		ConnectionsHandshakeErrors: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "test_connections_handshake_errors",
+		}),
+	}
 }

